File: LC/[0218] The Skyline Problem - v2.py
# from __future__ import annotations

from collections import Counter, defaultdict
from heapq import heappop, heappush
from typing import DefaultDict, List, Tuple

class Solution:
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        """
        Sweep line over x-coordinates.
        Use a max-heap of active heights (stored as negative values) and a Counter
        for lazy deletions of heights that have ended.
        """
        # Build events: at x, add list of starting heights and ending heights.
        # We'll encode starts as negative heights so they sort before ends at the same x.
        
        # python type annotation, and declaration events: events: DefaultDict[int, List[int]] = defaultdict(list)
        events: DefaultDict[int, List[int]] = defaultdict(list)
        for left, right, height in buildings:# process each building and construct all events (turning points)
            events[left].append(-height)  # start event
            events[right].append(height)  # end event

        # Active heights: max-heap via negative values
        heap: List[int] = [0]  # 0 means ground level (height 0) --- given a height, sort
        active = Counter({0: 1}) # count numf of heights at each ii

        result: List[List[int]] = [] # define return var
        prev_max = 0

        for x in sorted(events): #sort events and process each (for each ii --- building start/end point)
            # Apply all events at x.
            # Sort so starts (negative) are processed before ends (positive)
            # to avoid generating extra points at the same x.
            for h in sorted(events[x]): # for each height at this point
                if h < 0:  # start
                    height = -h
                    active[height] += 1
                    heappush(heap, -height) # maintain the max h
                else:      # end
                    height = h
                    active[height] -= 1 # count number of hight at this building point

            # Pop heap while the top height is no longer active (lazy deletion).
            while heap and active[-heap[0]] == 0:
                heappop(heap)

            curr_max = -heap[0]
            if curr_max != prev_max:
                result.append([x, curr_max])
                prev_max = curr_max

        return result


"""
# Problem Statements
* building[i]=[left, right, height]
       
         ----- height
         -----
0 ---left-----right

* no equal line

# Analysis
* N=10^4 -> nlogn
* building is already sorted
* we are given ranges -> we need to do range aggregation -> sweepline
1. collect events for a turning points: x:[+/-h]
2. for each turning points: could be 0 (start or end), could be H
"""

File: LC/[0218] The Skyline Problem - v2.java
import java.util.*;

class Solution {
    public List<List<Integer>> getSkyline(int[][] buildings) {
        // Each building [L, R, H] becomes two events:
        // (L, -H) start   -> negative height so it sorts before end at same x
        // (R, +H) end
        List<int[]> events = new ArrayList<>(buildings.length * 2);
        for (int[] b : buildings) {
            int L = b[0], R = b[1], H = b[2];
            events.add(new int[]{L, -H});
            events.add(new int[]{R,  H});
        }

        // Sort by:
        // 1) x ascending
        // 2) height ascending (so starts (-H) come before ends (+H) at same x;
        //    also taller starts processed first because -15 < -10)
        events.sort((a, b) -> {
            if (a[0] != b[0]) return Integer.compare(a[0], b[0]);
            return Integer.compare(a[1], b[1]);
        });

        // Max-heap for current heights (store positive heights, comparator reversed)
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        // Frequency map for lazy deletion / accurate counts
        Map<Integer, Integer> freq = new HashMap<>();

        // Initialize with ground height 0
        maxHeap.offer(0);
        freq.put(0, 1);

        List<List<Integer>> result = new ArrayList<>();
        int prevMax = 0;

        int i = 0;
        while (i < events.size()) {
            int x = events.get(i)[0];

            // Process all events at the same x together
            while (i < events.size() && events.get(i)[0] == x) {
                int h = events.get(i)[1];

                if (h < 0) { // start
                    int height = -h;
                    maxHeap.offer(height);
                    freq.put(height, freq.getOrDefault(height, 0) + 1);
                } else { // end
                    int height = h;
                    freq.put(height, freq.get(height) - 1);
                }
                i++;
            }

            // Clean heap top: remove heights that are no longer active
            while (!maxHeap.isEmpty() && freq.getOrDefault(maxHeap.peek(), 0) == 0) {
                maxHeap.poll();
            }

            int currMax = maxHeap.peek();
            if (currMax != prevMax) {
                result.add(Arrays.asList(x, currMax));
                prevMax = currMax;
            }
        }

        return result;
    }
}